### PR TITLE
Add *scan commands.

### PR DESCRIPTION
**Pull request by anparker**
Since #11 is a bit stalled, I made another implementation. Scan related commands will return cursor object, which will be used for iterations and cursor value handling.

### CODE REVIEW

File: lib/Mojo/Redis2.pm [hunk 1]
```diff
@@ -386,6 +387,18 @@ See L<http://redis.io/topics/pubsub> for details.
 The reverse of L</psubscribe>.
 See L<http://redis.io/topics/pubsub> for details.
 
+=head2 scan, hscan, sscan, zscan
+
+  $cur  = $self->scan(MATCH => 'namesoace*', COUNT => 15);
+  $hcur = $self->hscan('hash.key', MATCH => 'pref.*');
+  $scur = $self->sscan('set.key');
+  $zcur = $self->zscan('zset.key');
```
Comment (jhthorsen) [inline lib/Mojo/Redis2.pm]: I think the doc should just use $cur. Looks like each method returns different data.

File: lib/Mojo/Redis2/Cursor.pm [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+has _cursor => 0;
+
+sub again {
+  my $self = shift;
+  $self->_cursor(0);
+  delete $self->{_finished};
+  return $self;
+}
+
+sub command { shift->{_command}[0] }
```
Comment (jhthorsen) [inline lib/Mojo/Redis2/Cursor.pm]: In which usecase do you need to access the command() and key()?

File: lib/Mojo/Redis2/Cursor.pm [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+  unshift @_, 'MATCH' if $_[0] && !ref $_[0];
+  return $cur->slurp(@_);
+}
+
+sub new {
+  my $self    = shift->SUPER::new();
+  my $command = shift;
+  $self->_command($command eq 'SCAN' ? [$command] : [$command, shift]);
+  croak 'ERR Should not specify cursor value manualy.'
+    if looks_like_number $_[0];
```
Comment (jhthorsen) [inline lib/Mojo/Redis2/Cursor.pm]: No need for input check. The command should just fail in the redis if invalid input is given.

File: lib/Mojo/Redis2/Cursor.pm [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+  my $resp = $self->redis->_execute(
+    basic => @{$self->_command},
+    $self->_cursor, @$args, $cb ? ($wrapper) : ()
+  );
+  return $resp if $cb;
+  $cb = sub { $_[2] };
+  return $wrapper->(undef, '', $resp);
+}
+
+sub slurp {
```
Comment (jhthorsen) [inline lib/Mojo/Redis2/Cursor.pm]: What about calling it all() instead?

File: lib/Mojo/Redis2/Cursor.pm [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+  return $cur->slurp(@_);
+}
+
+sub new {
+  my $self    = shift->SUPER::new();
+  my $command = shift;
+  $self->_command($command eq 'SCAN' ? [$command] : [$command, shift]);
+  croak 'ERR Should not specify cursor value manualy.'
+    if looks_like_number $_[0];
+  $self->_args(\@_) if @_;
```
Comment (jhthorsen) [inline lib/Mojo/Redis2/Cursor.pm]: Looks like _args() and _command() should be merged to just one attribute command. We could then instantiate the object with a regular new():

    Mojo::Redis2::Cursor->new(command => [$op => @_]);

**Comment by jhthorsen** [comment]
I really appreciate this PR! It looks very nice. Please fix up / reply on my comments and I'll merge it when we have come to an agreement.

👍

**Comment by anparker** [comment]
`command` and `key` was originally an attributes. But I wasn't able to come up with a reason, why you may want to change them after object creation and made them read-only accessors. Should had removed them.

About merging command and args attributes, actually that was one of early implementation. :) Still can go with regular `new()` because redis guys for some reason put key before cursor.
And now you should specify 0 as initial cursor value, what feels somewhat pointless. At the other hand, now it's more consistent with other commands, that follow redis syntax without changing anything.

And I didn't change any existing code, not sure why travis started complaining about other tests.

**Comment by jhthorsen** [comment]
Awesome! Thanks for the contribution 👍

**Comment by anparker** [comment]
Can you fix an example in cursor class too? I forgot about it.
```
--- a/lib/Mojo/Redis2/Cursor.pm
+++ b/lib/Mojo/Redis2/Cursor.pm
@@ -119,7 +119,8 @@ Mojo::Redis2::Cursor - Cursor iterator for SCAN commands.
 
   my $redis = Mojo::Redis2->new();
   my $cursor
-    = Mojo::Redis2::Cursor->new('SCAN', match => 'mynamespace*')->redis($redis);
+    = Mojo::Redis2::Cursor->new(command => ['SCAN', 0, match => 'mynamespace*'])
+    ->redis($redis);
 
   # blocking
   while (my $r = $cursor->next) { say join "\n", @$r }
```

**Comment by jhthorsen** [comment]
I won't make a new release because of this, but the change is in the pipeline.

Thanks 👍

**Pull request merged**